checker/src/main/java/org/checkerframework/checker/resourceleak/MustCallConsistencyAnalyzer.java

```diff
@@ -0,0 +1,1708 @@
+   *     temporary variable (via a call to {@link
+   *     ResourceLeakAnnotatedTypeFactory#getTempVarForNode(Node)})
+   */
+  private void doGenKillForPseudoAssignment(
+      Node node,
+      Set<ImmutableSet<LocalVarWithTree>> obligations,
+      LocalVariableNode lhsVar,
+      Node rhs) {
+    // Replacements to eventually perform in obligations.  We keep this map to avoid a
+    // ConcurrentModificationException in the loop below.
+    Map<ImmutableSet<LocalVarWithTree>, ImmutableSet<LocalVarWithTree>> replacements =
+        new LinkedHashMap<>();
+    // construct lhsVarWithTreeToGen once outside the loop for efficiency
+    LocalVarWithTree lhsVarWithTreeToGen =
+        new LocalVarWithTree(new LocalVariable(lhsVar), node.getTree());
+    for (ImmutableSet<LocalVarWithTree> obligation : obligations) {
+      Set<LocalVarWithTree> kill = new LinkedHashSet<>();
+      // always kill the lhs var if present
+      addLocalVarWithTreeToSetIfPresent(obligation, lhsVar.getElement(), kill);
+      LocalVarWithTree gen = null;
+      // if rhs is a variable tracked in the obligation's resource alias set, gen the lhs
+      if (rhs instanceof LocalVariableNode) {
+        LocalVariableNode rhsVar = (LocalVariableNode) rhs;
+        for (LocalVarWithTree lvwt : obligation) {
+          if (lvwt.localVar.getElement().equals(rhsVar.getElement())) {
+            gen = lhsVarWithTreeToGen;
+            // We remove temp vars from tracking once they are assigned to another location.
+            if (typeFactory.isTempVar(rhsVar)) {
+              addLocalVarWithTreeToSetIfPresent(obligation, rhsVar.getElement(), kill);
+            }
+            break;
+          }
+        }
+      }
+      // Check if there is something to do before creating a new obligation, for efficiency.
+      if (kill.isEmpty() && gen == null) {
+        continue;
+      }
+      Set<LocalVarWithTree> newObligation = new LinkedHashSet<>(obligation);
+      newObligation.removeAll(kill);
+      if (gen != null) {
+        newObligation.add(gen);
+      }
+      if (newObligation.size() == 0) {
+        // We have killed the last reference to the resource; check the must-call obligation.
+        MustCallAnnotatedTypeFactory mcAtf =
+            typeFactory.getTypeFactoryOfSubchecker(MustCallChecker.class);
+        checkMustCall(
+            obligation,
+            typeFactory.getStoreBefore(node),
+            mcAtf.getStoreBefore(node),
+            "variable overwritten by assignment " + node.getTree());
+      }
+      replacements.put(obligation, ImmutableSet.copyOf(newObligation));
+    }
+    // Finally, update obligations according to the replacements.
+    for (Map.Entry<ImmutableSet<LocalVarWithTree>, ImmutableSet<LocalVarWithTree>> entry :
+        replacements.entrySet()) {
+      obligations.remove(entry.getKey());
+      if (!entry.getValue().isEmpty()) {
+        obligations.add(entry.getValue());
```

Comment:
Are we guaranteed that there can be no need to merge information?  That is, all the sets that represent all the obligations are disjoint -- no two entry values are related?

Reply:
I don't think there's any such guarantee - a single resource might be represented in more than one obligation (this is a consequence of the fact that we have a conservative must-alias analysis: there's not a guarantee that we've found all aliases, so there's no way to guarantee that we merge sets appropriately).

Reply:
Why do you ask? I don't think we ever imply in the documentation that such a guarantee exists (but if we did, I'd like to remove any reference to it).

Reply:
I ask because I'm concerned there might be a bug (or at least, I don't see why there isn't a bug).

I think that all the obligations/resource sets should be disjoint -- a given local/variable should not appear in multiple of them.  (Is that assumption correct?  Maybe I am confused about `obligations` itself.)

However, the `entry.getValue()` values are created one-by-one.  Each such obligation is disjoint from `obligations`, but there is no guarantee they are all disjoint from one another.  Thus, it seems possible that after this loop finishes and all the `entry.getValue()` values are added to `obligations`, the disjointness invariant of `obligations` might no longer hold.

Reply:
> I think that all the obligations/resource sets should be disjoint -- a given local/variable should not appear in multiple of them. (Is that assumption correct? Maybe I am confused about obligations itself.)

I don't think this is correct. As I said above, I don't think we ever make such a guarantee, nor do I understand why not doing so might be a bug.

That the same local variable might appear in multiple obligations makes sense: we're doing a conservative must-alias analysis, so there might be aliases that we're unaware of. So, conceptually, our algorithm has to work in the presence of "copies" (that is, aliases) of the same underlying resources appearing in multiple sets. At that point, I don't think it's any different to deal with multiple copies of the same local.

Generally, adding a new obligation related to some local is either a no-op or will add the possibility of additional errors being raised. In this sense, it's conservative, so I'm still confused about what, exactly, you think the problem with this might be. Maybe a call to discuss this would be useful?

Reply:
I agree that the same local variable can appear in multiple obligations in principle.  Consider:
```java
Socket s = null;
if (...) {
  s = new Socket(addr1,port1);
} else {
  s = new Socket(addr2,port2);
}
```
With this code, `s` will appear in two separate obligations after the `if`.  And, if `s` is not closed, we will report two errors, one for each obligation.